feat(api): fully connected volume tracking

pmoegenburg opened the pull request:
# Overview

This PR aims to consolidate well volume data from our existing relevant commands into a single source of truth in WellStore. These commands are LiquidProbe, LoadLiquid, Aspirate, and Dispense.

## volume estimation and height
Well volume and height are tracked in three parts:
- the probed height is a height gathered from a liquid probe since the last volume-affecting operation occurred; it is nulled out if you carry out a liquid operation on a well
- the tracked volume is a volume that started with a `LoadLiquid` and was then modified by liquid volume tracking from aspirate and dispense volumes
- the estimated volume is the same, but has been updated or started with a liquid height measurement that was converted into an estimated volume

The reason for having all three is that they have different accuracy characteristics, and we therefore use them for meniscus height estimation in an order of precedence:
- The most accurate is a probed height
- If a probed height isn't available, use a tracked volume that we turn into a height
- If that isn't available, use an estimated volume that we turn into a height

The latter two values are only accessible if the labware the well is in has internal geometry data. Even when we do, that geometry data is noisy and can be inaccurate, making those values suitable for problems like "I don't want to have to accurately measure how much is in this tube" and not for problems like "I want to minimize external droplet formation on my 1uL aspiration".

This PR implements all that, as well as setting the movement data for actually going to a well meniscus.

## volume offsets

A pretty important thing to consider for _static_ meniscus-relative movements is that when you're aspirating you don't want to move to where the meniscus _is_ but rather where it will be when the operation ends, so you don't leave your pipette high and dry. We signify this by allowing a "volume offset" in a location. This is like a height offset but is specified in uL or the special value "operationVolume". It will add an extra Z-offset to the positioning to account for a volume change in a well. This is calculated using the well's internal geometry and thus only available if the well has internal geometry. If the value is the string "operationVolume", which should be thought of as a sentinel value, we even calculate the volume offset from the volume you passed to `aspirate`; but specifying the value as a float is available for things like a `moveToWell` preceding an `aspirateInPlace`.

There's some gross internals because of history. One particularly gross one is the Location data type; this is an old friend indeed, and its friendship continues to pay dividends since we want to add something to it that is somewhat internal, and we have to do so in a really gross way.

## testing
We really need to test this in actually used protocols and this shouldn't be relied upon yet, but extensive test coverage should help.

pmoegenburg committed:
initial implementation for load_liquid volume-tracking

pmoegenburg committed:
Merge branch 'edge' into EXEC-281-volume-tracking

pmoegenburg committed:
updated naming

pmoegenburg committed:
updated implementation to include Aspirate and Dispense results

pmoegenburg committed:
eliminated previously used import

pmoegenburg committed:
added operations_since_measurement to LiquidHeightInfo

pmoegenburg committed:
refactor to use StateUpdate

pmoegenburg committed:
updated setting and getting liquid info and linted

pmoegenburg committed:
Merge branch 'edge' into EXEC-281-volume-tracking

pmoegenburg committed:
for less diff

pmoegenburg committed:
reverted typing changes

pmoegenburg committed:
1 more revert

pmoegenburg committed:
updated naming

pmoegenburg committed:
fixed WellStore dict handling

pmoegenburg committed:
renamed state update stuff and fixed tests

pmoegenburg committed:
refactored getting liquid values, fixed and added tests

pmoegenburg committed:
updated WellView get methods, fixed LiquidProbe to prevent errors, added tests for IncompleteLabwareDefinitionError

pmoegenburg committed:
removed comments

pmoegenburg committed:
added comments

pmoegenburg committed:
fixed lint error

pmoegenburg committed:
Merge branch 'edge' into EXEC-281-volume-tracking

sfoster1 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -765,6 +765,7 @@ def air_gap(
             raise RuntimeError("No previous Well cached to perform air gap")
         target = loc.labware.as_well().top(height)
         self.move_to(target, publish=False)
+        # add (is_liquid: bool = True) parameter to aspirate? Or deduce location is not in well in Aspirate command?
```
not quite sure what the note means?

sfoster1 commented on api/src/opentrons/protocol_engine/commands/aspirate.py:
```diff
@@ -140,6 +140,7 @@ async def execute(self, params: AspirateParams) -> _ExecuteReturn:
                 command_note_adder=self._command_note_adder,
             )
         except PipetteOverpressureError as e:
+            # can we get aspirated_amount_prior_to_error? If not, can we assume no liquid was removed from well? Ask Ryan
```
no we really can't, since pressure problems means the amount of liquid is unknown. It also cannot be assumed to be 0. we probably will have to null out the tracked volume for the well if this error occurs. Which means we may want to add measuring the liquid height to the error recovery

sfoster1 commented on api/src/opentrons/protocol_engine/commands/blow_out.py:
```diff
@@ -93,6 +93,7 @@ async def execute(self, params: BlowOutParams) -> _ExecuteReturn:
                 pipette_id=params.pipetteId, flow_rate=params.flowRate
             )
         except PipetteOverpressureError as e:
+            # can we get blown_out_amount_prior_to_error? If not, can we assume no liquid was added to well? Ask Ryan
```
notes comment. also no, see above

sfoster1 commented on api/src/opentrons/protocol_engine/commands/blow_out.py:
```diff
@@ -114,6 +115,7 @@ async def execute(self, params: BlowOutParams) -> _ExecuteReturn:
                 ),
             )
         else:
+            # if blow out over a well, can we get pipette tip liquid volume and add that to well liquid volume?
```
in theory yes, but the common case of blowing out is that you're doing so after a dispense with the known volume being 0. I think that this PR can avoid tracking blowouts at all and we can leave a note that if you blow-out without dispensing the volume won't be tracked.

sfoster1 commented on api/src/opentrons/protocol_engine/commands/dispense.py:
```diff
@@ -107,6 +107,7 @@ async def execute(self, params: DispenseParams) -> _ExecuteReturn:
                 push_out=params.pushOut,
             )
         except PipetteOverpressureError as e:
+            # can we get dispensed_amount_prior_to_error? If not, can we assume no liquid was added to well? Ask Ryan
```
see above

sfoster1 commented on api/src/opentrons/protocol_engine/state/state_summary.py:
```diff
@@ -30,5 +32,7 @@ class StateSummary(BaseModel):
     completedAt: Optional[datetime]
     liquids: List[Liquid] = Field(default_factory=list)
-    wells: List[LiquidHeightSummary] = Field(default_factory=list)
+    wells: List[
```
this doesn't seem right. this would represent a list like this: `[LoadedVolumeSummary(...), LoadedVolumeSummary(...), ProbedVolumeSummary(...), LoadedVolumeSummary(...), ProbedHeightSummary(...), ...]` with no well identification or anything. This probably needs some more structure or something, like a dataclass that includes all three summaries, a well identifier, and a labware id

sfoster1 commented on api/src/opentrons/protocol_engine/state/wells.py:
```diff
@@ -1,25 +1,28 @@
     """State of all wells."""
 
-    measured_liquid_heights: Dict[str, Dict[str, LiquidHeightInfo]]
+    loaded_volumes: Dict[str, Dict[str, LoadedVolumeInfo]]
```
these dicts are `{labware_id: { well_name: info } }` right? maybe we should have that in a comment since we can't capture it as a type, or honestly maybe we should have `LabwareId = str` somewhere above just so you can see what the strings are when you look at them.

sfoster1 commented on api/src/opentrons/protocol_engine/state/wells.py:
```diff
@@ -29,41 +32,87 @@ class WellStore(HasState[WellState], HandlesActions):
+            self._handle_liquid_probed_update(state_update)
+            self._handle_liquid_operated_update(state_update)
+
+    def _handle_liquid_loaded_update(
```
this is really more of a personal preference thing, but what if this function was `def _handle_liquid_loaded_update(self, state_update: LiquidLoadedUpdate) -> None`, and we put that first `if` statement at the call site? Then this function has less indentation and doesn't have the weird thing where it does nothing if there's a missing if statement. The place we call this can also take advantage of pattern matching to call this without even having an explicit `if state_update.liquid_loaded != update_types.NO_CHANGE`.

pmoegenburg replied:
Anddd I don't see a way for match to make code cleaner, and moving the if statements to `handle_action()` results in lint errors in `_handle_xyz()` that would require asserts/more code, so keeping as is.

sfoster1 reviewed (changes requested):
Couple small things to change and some answers to the questions. Also I think we're missing handling for the in place liquid handling commands? Structure looks great!

caila-marashaj commented on api/src/opentrons/protocol_engine/commands/liquid_probe.py:
```diff
@@ -220,6 +228,21 @@ async def execute(self, params: _CommonParams) -> _LiquidProbeExecuteReturn:
             )
         else:
+            try:
+                well_volume = self._state_view.geometry.get_well_volume_at_height(
```
does this `well_volume` get returned somewhere?

pmoegenburg replied:
it's an arg to the `state_update.set_liquid_probed()` call just below

caila-marashaj commented on api/src/opentrons/protocol_engine/state/geometry.py:
```diff
@@ -1401,15 +1401,29 @@ def get_meniscus_height(
+                well_name=well_name,
+                volume=loaded_volume_info.volume,
+            )
+        elif probed_volume_info is not None and probed_volume_info.volume is not None:
```
Does `probed_volume` mean we probed a height during a previous liquid operation, and modified the volume found from that probed height?

pmoegenburg replied:
Yup, we create probed_height_info and probed_volume_info when we LiquidProbe, then clear probed_height_info and update probed_volume_info whenever we Aspirate/Dispense. So if there hasn't been a LoadLiquid and there's been a LiquidProbe followed by an Aspirate/Dispense, we'll use probed_volume_info

pmoegenburg commented:
> Couple small things to change and some answers to the questions. Also I think we're missing handling for the in place liquid handling commands? Structure looks great!

We only call AspirateInPlace from one place (InstrumentCore.aspirate()), and only call it in that place if the location is not a well, correct? Same with DispenseInPlace. So updating WellStore from AspirateInPlace and DispenseInPlace isn't needed, correct?

caila-marashaj commented on api/src/opentrons/protocol_engine/state/wells.py:
```diff
@@ -29,41 +32,87 @@ class WellStore(HasState[WellState], HandlesActions):
+                and well_name in self._state.loaded_volumes[labware_id]
+            ):
+                prev_loaded_vol_info = self._state.loaded_volumes[labware_id][well_name]
+                assert prev_loaded_vol_info.volume is not None
```
If we start with a loaded volume, and then do some operation on it and have a new volume, should this new resulting volume still be considered a `loaded_volume`, or is it less reliable enough that it should be considered something different?

pmoegenburg replied:
I think the operations_since_load/probe will get factored into the offset from meniscus. So if you specify aspirate.meniscus, we'll make sure the WellOffset is at least like (0.5mm * operations_since_load/probe) below the meniscus

caila-marashaj commented on api/src/opentrons/protocol_engine/state/wells.py:
```diff
@@ -79,51 +128,71 @@ def __init__(self, state: WellState) -> None:
+            Union[ProbedHeightSummary, ProbedVolumeSummary, LoadedVolumeSummary]
+        ] = []
+        for lv_labware, lv_wells in self._state.loaded_volumes.items():
+            for lv_well, lvi in lv_wells.items():
```
What is `lvi` here?

pmoegenburg replied:
LiquidVolumeInfo

pmoegenburg committed:
updates from Seth's PR review

pmoegenburg committed:
Merge branch 'edge' into EXEC-281-volume-tracking

pmoegenburg committed:
added LiquidOperatedUpdate to aspirate/dispense_in_place

pmoegenburg committed:
removed/addressed comment

sfoster1 committed:
in place liquid failure handling

sfoster1 committed:
yes, but which one

sfoster1 committed:
whoops

sfoster1 committed:
this can definitely be none it means we didn't have schema 3 labware

sfoster1 committed:
let's at least do some logging

sfoster1 committed:
format

sfoster1 reviewed (approved):
clearing previous review

sfoster1 committed:
format again

sfoster1 committed:
fix: do not force operationvolume offset

This is not the right thing to be doing, we want callers to be able to
say what they want.

sfoster1 committed:
fix: set operationVolume in the core

SyntaxColoring commented on api/src/opentrons/protocol_engine/types.py:
```diff
@@ -355,20 +355,41 @@ class CurrentWell:
+
+class WellInfoSummary(BaseModel):
+    """Payload for a well's liquid info in StateSummary."""
 
     labware_id: str
     well_name: str
-    height: float
-    last_measured: datetime
+    loaded_volume: Optional[float] = None
+    last_loaded: Optional[datetime] = None
+    operations_since_load: Optional[int] = None
+    probed_height: Optional[float] = None
+    probed_volume: Optional[float] = None
+    last_probed: Optional[datetime] = None
+    operations_since_probe: Optional[int] = None
```
Do we really want to expose all of these `WellInfoSummary` fields over HTTP and store them in the database? Is there a client need for all of these fields?

sfoster1 replied:
I eagerly want to expose these fields over HTTP because I think this system is better when it exposes more information, though storing it in the database is a bit icky. There isn't a currently planned in the app use for them but there is a use for me wanting to see what the information is.

ryanthecoder reviewed (approved):
LGTM

ryanthecoder commented on api/src/opentrons/protocol_engine/state/wells.py:
```diff
@@ -29,41 +34,94 @@ class WellStore(HasState[WellState], HandlesActions):
-                well_name=action.error.private.well_name,
-                height=None,
-                time=action.failed_at,
+        state_update = get_state_update(action)
```
```suggestion
        state_update = get_state_update.get_state_update(action)
```

TamarZanzouri commented on api/src/opentrons/protocol_engine/execution/queue_worker.py:
```diff
@@ -69,7 +69,11 @@ async def join(self) -> None:
-            await self._command_executor.execute(command_id=command_id)
+            try:
+                await self._command_executor.execute(command_id=command_id)
+            except BaseException:
```
is this intentional?

sfoster1 replied:
yes. it could also be a separate change probably but not having this is why tests would mysteriously hang sometimes

sfoster1 committed:
imports r hard

sfoster1 committed:
really hard, i swear

TamarZanzouri commented on api/src/opentrons/protocol_engine/state/wells.py:
```diff
@@ -79,51 +138,100 @@ def __init__(self, state: WellState) -> None:
-            )
-        except KeyError:
-            return False
+    ) -> Tuple[
```
a bit worried about returning a tuple with optional values?

sfoster1 committed:
Merge branch 'edge' into EXEC-281-volume-tracking

SyntaxColoring commented on api/src/opentrons/protocol_engine/types.py:
```diff
@@ -355,20 +355,41 @@ class CurrentWell:
-    height: float
-    last_measured: datetime
+    loaded_volume: Optional[float] = None
+    last_loaded: Optional[datetime] = None
```
This doesn't need to change for this PR, but I think keeping track of, and exposing, these `datetime`s is an uncomfortable middle ground. I think either:

* We want to figure out a more structured, machine-readable way for clients to trace the provenance of these measurements. For example, pointing back to the original loading command `id`. (Peter tried this but ran into architectural problems: command implementations do not know the ID of the command they're running in, and I'm not sure we want them to know.)
* Or, we don't actually care about this and we can delete it. Maybe we'd replace it with `notes` if we still want something human-readable for debugging.

sfoster1 replied:
I agree, and I think I'll remove them for now until we can add the IDs there

sfoster1 committed:
merge fixups

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/update_types.py:
```diff
@@ -330,3 +366,46 @@ def mark_tips_as_used(
+            labware_id=labware_id,
+            well_name=well_name,
+            volume=volume,
+        )
```
Some nitpicks:

* Can we rename `volume` to something like `volume_delta` or `volume_added_to_well`, to make it clearer that it doesn't just set the new volume to exactly what you supplied?
* Instead of `None`, should we use the `CLEAR` sentinel to make it more obvious what happens when you supply that?
* Can we remove the `= None` (or `= CLEAR`) argument default? If someone forgets to supply a value, deleting the volume is probably not a safe default; the caller should be forced to write what they want to do explicitly.

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/update_types.py:
```diff
@@ -330,3 +366,46 @@ def mark_tips_as_used(
+        well_name: str,
+        last_probed: datetime,
+        height: typing.Optional[float] = None,
+        volume: typing.Optional[float] = None,
```
Same nitpicks about switching from `None` to `CLEAR` and about not having default argument values.

sfoster1 committed:
use a dataclass instead of atuple

sfoster1 committed:
oops

sfoster1 committed:
lint

TamarZanzouri commented on api/src/opentrons/protocol_engine/state/wells.py:
```diff
@@ -79,51 +137,100 @@ def __init__(self, state: WellState) -> None:
+                    operations_since_load=lvi.operations_since_load,
+                )
+                summaries.append(whi)
+        summaries = self._append_summaries_with_probed_heights(summaries=summaries)
+        summaries = self._append_summaries_with_probed_volumes(summaries=summaries)
```
this is a bit confusing to me. can we not return anything? the list is mutable

TamarZanzouri reviewed (changes requested):
a few requests. thank you!

SyntaxColoring reviewed (commented):
This all looks reasonable to me so far. I'm catching up and working my way through the `wells.py` stuff now. Meanwhile, here are some thoughts for the other parts.

sfoster1 committed:
review: better summarization

bit more readable, bit less datetime

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/aspirate_in_place.py:
```diff
@@ -100,6 +106,16 @@ async def execute(self, params: AspirateInPlaceParams) -> _ExecuteReturn:
                 command_note_adder=self._command_note_adder,
             )
         except PipetteOverpressureError as e:
+            # TODO(pbm, 10-24-24): if location is a well, get new tip and LiquidProbe in error recovery to reestablish well liquid level
```
We might want to remove these `# todo` comments in the command implementations because this seems like a big cross-stack thing with designs and everything; it's not an implementation detail that we can "just do" on the backend.

sfoster1 replied:
that's fair

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/geometry.py:
```diff
@@ -1401,15 +1401,36 @@ def get_meniscus_height(
+                well_name=well_name,
+                volume=well_liquid.loaded_volume.volume,
+            )
+        elif (
+            well_liquid.probed_volume is not None
+            and well_liquid.probed_volume.volume is not None
+        ):
+            return self.get_well_height_at_volume(
+                labware_id=labware_id,
+                well_name=well_name,
+                volume=well_liquid.probed_volume.volume,
```
This ladder looks like:

1. Return the last probed height, directly.
2. Or return the height computed from the loaded volume, where the "loaded volume" is either from a `load_liquid()`, or from adding up `aspirate()`s and `dispense()`s.
3. (these highlighted lines) Or return the height computed from the last probed volume.

But in part 3, wasn't the "last probed volume" computed from the last probed height? So isn't part 3 exactly equivalent to part 1? Is part 3 doing anything?

sfoster1 replied:
no, because the key with (1) is that the last probed height is available if and only if you haven't since touched the volume, whereas (3) is always available

sfoster1 replied:
the use cases for (1) look like, "do a probe immediately before every aspirate for maximum accuracy". the use case for (3) looks like "i have this one tube of master mix I fill up when it's getting empty and I don't care how much is in it beyond 'enough', so I'll probe it once at the beginning of the protocol and let the robot track volume from there"

SyntaxColoring replied:
> no, because the key with (1) is that the last probed height is available if and only if you haven't since touched the volume, whereas (3) is always available

Ohh I see, thanks. So when we touch the volume, the element in `probed_heights` is deleted, but the element in `probed_volumes` is not.

-----

This certainly doesn't have to change in this PR, but what do you think about simplifying this by expressing it in `WellState` as attributes like:

* `current_baseline`, which stores, per-well, either:
    * The most recent volume from `loadLiquid` (even if there were a bunch of liquid handling ops since then)
    * Or the most recent height from `probeLiquid` (even if there were a bunch of liquid handling ops since then)
* `volume_change_since_baseline`, which stores, per-well, the aspirate+dispense+etc. volume delta since then.

So then, going through the three modes, using your descriptions from Slack:

1. *mode 1: meniscus relative aspiration from a recent LLD height estimation (ie. probe then aspirate). this is the "gold standard"...*

   This is when `current_baseline` is a `liquidProbe` height and `volume_change_since_baseline == 0`.

2. *mode 2: meniscus relative aspiration from volume-based height estimation, from a volume entered in the protocol with load_liquid and tracked through pipette actions. less accurate than immediate LLD...*

   This is when `current_baseline` is a `loadLiquid` volume.

3. *mode 3: meniscus relative aspriation from volume-based height estimation from height-based volume estimation, i.e. you got your initial volume from an LLD pass. this is the worst because you suffer those errors twice...*

   This is when `current_baseline` is a `liquidProbe` height and `volume_change_since_baseline != 0`.

I guess I'm shooting for `WellState` attributes to represent exactly what we know, in plain terms, without any editorializing. Then the `GeometryView` methods can implement the entirety of our business logic, with the cascading accuracy thing, based on those ground truths. Right now, I feel like the business logic is leaking into `WellState` and muddying what those attributes are. (`probed_heights` and `probed_volumes` differ in more ways than just "one is in height units and one is in volume units.")

sfoster1 replied:
I see what you're getting at. My bias is towards doing that but also storing separate data separately. What I think that version would be is
- `loaded_volume_baseline`: the last `load_volume`
- `probed_height_baseline`: the last `probe_liquid`, whenever it was
- `volume_change_since_loaded_volume_baseline`
- `volume_change_since_probed_volume_baseline` (we need both because you might have probed after some liquid operations on a loaded baseline)
- something like `immediate_probed_height`. This could also just be `probed_height_baseline` if we're tracking how many operations it's been since then, but I think it's much nicer and inspectable to be able to look at the data and pick between the entries, rather than follow a chain of logic.

SyntaxColoring reviewed (approved):
Looks good to merge if all comments are resolved to your satisfaction. I agree that we should get this in, get some geometry definitions in, and play around with it to see what breaks.

sfoster1 committed:
clear instead of None

sfoster1 committed:
remove some todos

sfoster1 committed:
volume_added

TamarZanzouri reviewed (approved):
Thank you for addressing the changes! Lgtm!
